### DISCUSSION TITLE
Scene Reference serialization

### DISCUSSION THREAD

**Discussion by Rava4U** [starikcetin/Eflatun.SceneReference#26] [Q&A] [answered]
Hi!
In a game I'm trying to develop (as a total beginner, by the way, so please be patient with me:-) ) I'm using the JohannesMP's version of this Scene Reference tool, which I'm finding extremely helpful. However, I came across a small personal issue: it looks like SceneReference is not serializable, or at least trying to create a save file containing Scene Reference through the binary formatter causes a Serialization Exception error.
My questions are:
- is JohannesMP's SceneReference serializable? I couldn't find the answer by myself and, if I understood correctly, that version of the project is no longer supported, that's why I came here!
- is THIS Scene Reference tool serializable, alternatively? It would be a great help if it was (otherwise, I have a suitable, even if less practical, plan B for what I need to do).
Any help is greatly welcome!

Andrea

PS: I'm also new to GitHub, so forgive me if I'm asking things the wrong way or in the wrong place

**Comment by starikcetin** [comment]
Hello,

Don't worry, you asked your question in the right place. Can you give me a little more detail on exactly what you are trying to do? Are we talking about Unity's binary serialization or something else?

Also, can you kindly share the full exception you are getting, along with the steps to reproduce it?

**Reply by Rava4U** [reply]
Let's start by admitting that I know nothing about serialization... what I'm doing in my project is using the, let's call it, "FileStream / BinaryFormatter" method to save and load progresses, so yes, I guess we're talking about Unity's binary serialization.

My game is level-based and I'm building each level in a different scene; levels are going to be accessed from a main world, so in the beginning I was manually assigning level names as strings to the levels' "doors". But even in my inexperience I knew it wasn't very practical... then I found this Scene Reference tool and I'm now assigning Scenes assets (instead of strings) to the "doors".
The saving system I set up, however, is still based on levels' names as strings. It works fine, but I realised it would be difficult (if not impossible) to update a level (maybe just to correct a misspelling) without breaking the game or overwriting player's progresses. So, I came up with a solution that requires introducing in my "Savedata" class a Dictionary<SceneReference, string>; if I got it right, having a reference to the Scene asset would uniquely identify a level despite any changes to its name and/or location, which would be enough for me.
Can't say if this solution is elegant or not, however I couldn't test it because, at first saving attempt, the game gave me this Serialization exception, the same that you get when you try to serialize Vector3, for example. I'll add that the Savedata class has its Serializable attribute (as every tutorial taught me :-) )
I'm afraid I can't give you the exact exception because I deleted the new codes and went back to the previous backup while re-thinking the problem (yes, I guess I was too rushed in doing that), but I can rewrite that code part to reproduce the error, if it can be of help.
Thanks a lot! I hope I've been clear enough!

**Comment by starikcetin** [comment, marked as answer]
OK, I think I understand what you are trying to do. I will test myself to see what problems binary serialization would have. As a sidenote, to prevent confusion, this is not the binary serialization of Unity, this is your custom binary serialization solution.

In the meantime, however, I think you have a much more elegant way to tackle this issue. Instead of serializing the whole `SceneReference` to your save file, you can just serialize the GUID of the scene. You can get it via `mySceneReference.AssetGuidHex`. Then, in the runtime, [you can access the scene GUID to scene path map directly](https://github.com/starikcetin/Eflatun.SceneReference#accessing-the-scene-guid-to-path-map-directly), and load the scene from its path.

**Reply by Rava4U** [reply]
I've re-written my solution referring to the GUID instead of the whole SceneReference, as you suggested, tested it a bit and it looks like it works exactly like I meant it to! Wonderful!

Many thanks for your help (and for your patience towards this... toddler developer! XD)

**Reply by starikcetin** [reply]
Everyone starts somewhere. You are much welcome. :)

**Comment by starikcetin** [comment]
Support for [custom serialization](https://github.com/starikcetin/Eflatun.SceneReference#custom-serialization) is included in [1.4.0](https://github.com/starikcetin/Eflatun.SceneReference/releases/tag/1.4.0).
You can now serialize a `SceneReference` using `BinaryFormatter`, `Newtonsoft.Json`, and alike.
